samza-core/src/main/java/org/apache/samza/operators/triggers/TimeSinceLastMessageTriggerImpl.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.operators.triggers;
+
+import org.apache.samza.operators.data.MessageEnvelope;
+
+/**
+ * Implementation class for a {@link TimeSinceLastMessageTrigger}
+ * @param <M>
+ */
+public class TimeSinceLastMessageTriggerImpl<M extends MessageEnvelope> extends TriggerImpl<M> {
+
+  private final TimeSinceLastMessageTrigger trigger;
+  private final long durationMs;
+  private long callbackTime = Integer.MIN_VALUE;
+  private Cancellable latestFuture = null;
+
+  public TimeSinceLastMessageTriggerImpl(TimeSinceLastMessageTrigger<M> trigger, TriggerContext context, TriggerCallbackHandler handler) {
+    super(context, handler);
+    this.trigger = trigger;
+    this.durationMs = trigger.getDuration().toMillis();
+  }
+
+
+  @Override
+  public void onMessage(M message) {
+
+    long currTime = System.currentTimeMillis();
+
+    if (currTime < callbackTime && latestFuture != null) {
+      latestFuture.cancel();
+    }
+
+    callbackTime = currTime + durationMs;
+    Runnable runnable = () -> {
+      handler.onTrigger(this, context.getWindowKey());
+    };
+
+    latestFuture = context.scheduleCallback(runnable, callbackTime);
+  }
+
+  @Override
+  public void onCancel() {
+    latestFuture.cancel();
```

Comment:
Can onCancel be called before before onMessage? If so, then it seems we need a guard in several of these classes. If not, we should document this in TriggerImpl so it is explicit.

---

One other thought on cancellation: it seems that in all cases I've seen so far that cancellation involves not doing something in onMessage or cancelling a timer. I wonder if we couldn't simplify the duplication happening for these two cases by handling cancellation a level above this: if cancel is invoked then we no longer call onMessage and any scheduled callbacks (assuming there is one given my understanding of the design) would also be cancelled immediately.

This restricts what can be done in cancellation. We could still provide the onCancel hook if needed for other cleanup. Alternatively if this actually complicates the implementation we should certainly drop the idea.